server/preprocessing/other-scripts/vis_layout.R

```diff
@@ -18,130 +18,189 @@ vis_layout <- function(text, metadata, max_clusters=15, maxit=500, mindim=2, max
 
-replace_keywords_if_empty <- function(corpus, metadata) {
-  
-  remove_alone_numbers <- content_transformer(function(x)
-    gsub('\\b\\d+\\s','', x))
-  
-  corpus <- tm_map(corpus, remove_alone_numbers)
-  
-  dtm = DocumentTermMatrix(corpus)
-  
-  i = 1
-  
-  for(i in 1:nrow(metadata)) {
-    if (metadata$subject[i] == "" || metadata$subject[i] == "n/a") {
-      freq_terms = as.matrix(dtm[i,])
-      freq_terms_sorted = sort(colSums(freq_terms), decreasing=TRUE)
-      top_terms = head(freq_terms_sorted, 10)
-      
-      metadata$subject[i] = paste0(names(top_terms), collapse=";")
-    }
-  }
-  
   return(metadata)
-  
+
 }
 
 normalize_matrix <- function(tdm_matrix, method = "cosine") {
   distance_matrix_2 <- as.matrix(proxy::dist(tdm_matrix, method))
   distance_matrix = as.dist(distance_matrix_2)
-  
+
   return(distance_matrix)
 }
 
 create_clusters <- function(distance_matrix, max_clusters=-1, method="ward.D") {
   # Perform clustering, use elbow to determine a good number of clusters
   css_cluster <- css.hclust(distance_matrix, hclust.FUN.MoreArgs=list(method="ward.D"))
-  cut_off = elbow.batch(css_cluster)
-  
+  cut_off <<- elbow.batch(css_cluster)
+
   num_clusters = cut_off$k
-  
+
   if(max_clusters > -1 && num_clusters > max_clusters) {
     num_clusters = MAX_CLUSTERS
   }
-  
+
   meta_cluster = attr(css_cluster,"meta")
   cluster = meta_cluster$hclust.obj
   labels = labels(distance_matrix)
   groups <- cutree(cluster, k=num_clusters)
-  
+
   if(debug == TRUE) {
     # Plot result of clustering to PDF file
     pdf("clustering.pdf", width=19, height=12)
     plot(cluster, labels=metadata$title, cex=0.6)
     rect.hclust(cluster, k=num_clusters, border="red")
     dev.off()
   }
-  
+
   clusters = list("labels"=labels, "cluster"=cluster, "groups"=groups, "num_clusters"=num_clusters)
   return(clusters)
-  
+
 }
 
 create_ordination <- function(distance_matrix, mindim=2, maxdim=2, maxit=500) {
 
   # Perform non-metric multidimensional scaling
-  nm = par.nmds(distance_matrix, mindim=mindim, maxdim=maxdim, maxit=maxit)
+  nm <<- par.nmds(distance_matrix, mindim=mindim, maxdim=maxdim, maxit=maxit)
   nm.nmin = nmds.min(nm)
-  
+
   if(debug == TRUE) {
     # Plot results from multidimensional scaling, highlight clusters with symbols
     pdf("mds.pdf")
     plot(nm.nmin, pch=groups)
     dev.off()
   }
-  
+
   return(nm.nmin)
 }
 
+
+SplitTokenizer <- function(x) {
+  tokens = unlist(lapply(strsplit(words(x), split=";"), paste), use.names = FALSE)
+  return(tokens)
+}
+
+
+create_cluster_names <- function(clusters, metadata_full_subjects, weightingspec, top_n) {
+  subjectlist = list()
+  for (k in seq(1, clusters$num_clusters)) {
+    group = c(names(clusters$groups[clusters$groups == k]))
+    matches = which(metadata_full_subjects$id%in%group)
+
+    titles =  metadata_full_subjects$title[c(matches)]
+    subjects = metadata_full_subjects$subject[c(matches)]
+    texts = mapply(paste, titles, subjects)
```

Comment:
This was based on our discussion with @brembs that keywords may not be totally representative of content, as they could be used similar to search engine optimization. I added paper titles as a source of keywords more accurately describing the content to be expected.